### DISCUSSION TITLE
Deprecated `jax.core.ConcreteArray`

### DISCUSSION THREAD

**Discussion by ASKabalan** [jax-ml/jax#25337] [Q&A] [answered]
Hello JAX team,

You have deprecated  `jax.core.ConcreteArray`
What can we use as replacement now?

Thanks

**Comment by jakevdp** [comment]
Hi - `jax.core.ConcreteArray` was an internal implementation detail that's no longer used. The abstract value type hierarchy was simplified in v0.4.36, and tracer objects that previously used `ConcreteArray` now use `ShapedArray` instead.

Can you say more about your use-case? You shouldn't have been directly accessing `ConcreteArray` at all, so it's hard to say what the replacement would be without knowing more about your use-case. Thanks!

**Reply by ASKabalan** [reply]
Thanks for you answer.

It was just to check if a tracer is concrete or not but [is_concrete](https://github.com/jax-ml/jax/blob/main/jax/core.py#L86) does the trick for me and it is exposed by you guys, so I assume that it is safe to use

**Reply by jakevdp** [reply, marked as answer]
APIs in `jax.core` are not considered public, and may be changed in future releases without warning (see the [API compatibility policy](https://jax.readthedocs.io/en/latest/api_compatibility.html)).

In general, branching logic on concreteness of tracers is probably not a good idea, because unless done *very* carefully it could lead to unintended effects (such as JIT compilation leading to differing outputs).

Can you say more about what you're trying to accomplish with this check?
